tracks/python/exercises/saddle-points/mentoring.md

```diff
@@ -0,0 +1,195 @@
+`IndexError` from trying to index into an empty list.
+
+```python
+# Best solution2
+if not matrix:
+    return []
+if not all(len(row) == len(matrix[0]) for row in matrix):
+    raise ValueError("irregular matrix")```
+# or:
+if any(len(row) != len(matrix[0]) for row in matrix):
+    raise ValueError("irregular matrix")```
+
+        
+# Acceptable solution
+if len(matrix) == 0:
+  return []
+row_length = len(matrix[0])
+for row in matrix:
+  if len(row) != row_length:
+    raise ValueError("irregular matrix")
+ ```
+
+This might be a good place to introduce concepts like list 
+comprehensions and the `all()` or `any()` function, which 
+are easier to use here but will be helpful for the main
+problem.
+
+### Iterating over the values
+
+There are various ways of iterating over the values, but nested 
+`for` loops are the most straightforward. Since the column and 
+row indicies are required along with the values, this is a good 
+place to introduce the `enumerate()` function. Alternatively, 
+students might use `range()` and `len()` functions and retrive
+the values by index.
+
+```python
+for row_num, row in enumerate(matrix):
+    for column_num in range(len(row)):
+```
+
+### Checking for saddle points
+
+Students must check that the value is the highest value in that
+row and the lowest value in the column. This is a good place to
+introduce the `max()` and `min()` functions which take an iterable
+and return the highest or lowest value. Alternatively students might 
+use the `<` and `>` operators and compare the value to each value
+in that row and column. This appreach is valid but generally results
+is much more verbose code. 
+
+Another challenge here is getting the column. At this stage most 
+students should be ready to use a comprehension which is the best way, 
+but students who aren't familar with comprehensions might use an 
+additional `for` loop instead. To make it simpler, suggest creating 
+a separate function for getting the column. 
+
+Students will often write more complicated checks than required. 
+Once we've extracted the row and column that we want to check, all
+we need to do is check that the `max` of the row is equal to the `min`
+of the column. Guide them to think about what we're checking for and
+suggest that they experiment with different checks to see what works.
+
+```
+# Suggested solution
+if max(row) == min(row[column_num] for row in matrix):
+
+# Alternative solution:
+def get_column(matrix, col_index):
+  return row[col_index] for row in matrix
+if max(row) == min(get_column(matrix, col_index):
+```
+
+### Returning the result
+
+The task requires returning a `list` of `dict` objects, one for each 
+saddle point, with the keys "row" and "column" and the values being
+the one-based index (i.e. the location, not the value, of the saddle 
+points).
+
+```
+result.append({"row": row_num + 1, "column": column_num + 1})
+    return result
+```
+
+There are various ways to structure the code for this problem; 
```

Comment:
These four paragraphs are good suggestions for code reviews in general but are not guides for this exercuse.